src/nevermined/Assets.ts

```diff
@@ -115,16 +146,47 @@ export class Assets extends Instantiable {
     }
 
     /**
-     * Returns a DDO by DID.
+     * Returns a DDO by DID. Depending of the resolution policy it prioritize the Metadata API or Immutable urls.
      * @param did - Decentralized ID.
+     * @param policy - It specifies the resolve policy to apply. It allows to select that priorities during the asset resolution via Metadata API or Immutable URLs (IPFS, Filecoin, etc)
      * @returns {@link DDO}
      */
-    public async resolve(did: string): Promise<DDO> {
-        const { serviceEndpoint } =
+    public async resolve(did: string, policy: DIDResolvePolicy = DIDResolvePolicy.ImmutableFirst): Promise<DDO> {
+        const { serviceEndpoint, immutableUrl } =
             await this.nevermined.keeper.didRegistry.getAttributesByDid(did)
-        return this.nevermined.metadata.retrieveDDOByUrl(serviceEndpoint)
+
+        const immutableUrlValid = immutableUrl && immutableUrl.length > 10
+
+        if (policy === DIDResolvePolicy.ImmutableFirst && !immutableUrlValid)
+            policy = DIDResolvePolicy.OnlyMetadataAPI
+
+        if (policy === DIDResolvePolicy.OnlyMetadataAPI) {
+            return this.nevermined.metadata.retrieveDDOByUrl(serviceEndpoint)
+
+        } else if (policy === DIDResolvePolicy.OnlyImmutable && immutableUrlValid) {
+            return this.nevermined.metadata.retrieveDDOFromExternalBackend(immutableUrl)
+
+        } else if (policy === DIDResolvePolicy.MetadataAPIFirst)    {
+            try {
+                return this.nevermined.metadata.retrieveDDOByUrl(serviceEndpoint)
+            } catch (error) {
+                if (immutableUrlValid)
+                    return this.nevermined.metadata.retrieveDDOFromExternalBackend(immutableUrl)
+                throw new AssetError(`Metadata of asset with did ${did} not found`)
+            }
+
+        } else if (policy === DIDResolvePolicy.ImmutableFirst)    {
+            try {
+                return this.nevermined.metadata.retrieveDDOFromExternalBackend(immutableUrl)
+            } catch (error) {
+                return this.nevermined.metadata.retrieveDDOByUrl(serviceEndpoint)
+            }
+        }
```

Comment:
I don't think we should be changing policies on the fly. We should just throw an error instead of defaulting to a different policy

Reply:
refactored